OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs

```diff
@@ -65,38 +65,35 @@ enum Edges : byte
 			All = Top | Right | Bottom | Left
 		}
 
-		struct ShroudTile
+		struct TileInfo
 		{
 			public readonly float2 ScreenPosition;
 			public readonly byte Variant;
 
-			public Sprite Fog;
-			public Sprite Shroud;
-
-			public ShroudTile(float2 screenPosition, byte variant)
+			public TileInfo(float2 screenPosition, byte variant)
 			{
 				ScreenPosition = screenPosition;
 				Variant = variant;
-
-				Fog = null;
-				Shroud = null;
 			}
 		}
 
 		readonly ShroudRendererInfo info;
-		readonly Sprite[] shroudSprites, fogSprites;
-		readonly byte[] spriteMap;
-		readonly CellLayer<ShroudTile> tiles;
-		readonly byte variantStride;
 		readonly Map map;
 		readonly Edges notVisibleEdges;
+		readonly byte variantStride;
+		readonly byte[] edgesToSpriteIndexOffset;
 
-		bool clearedForNullShroud;
-		int lastShroudHash;
-		CellRegion updatedRegion;
+		readonly CellLayer<TileInfo> tileInfos;
+		readonly CellLayer<bool> shroudDirty;
 
+		readonly Vertex[] fogVertices, shroudVertices;
+		readonly Sprite[] fogSprites, shroudSprites;
+		readonly CellLayer<Sprite> fogSpriteLayer, shroudSpriteLayer;
 		PaletteReference fogPalette, shroudPalette;
```

Comment:
I'm grouping this here because they're semantically related. Everything here is a fog/shroud combo. I'd rather do this that group by readonly-ness unless there's a particular need.